### PR TITLE
improve stability of density estimates near boundary

### PR DESCRIPTION
**Pull request by tnagler**

### CODE REVIEW

File: R/kde1d_methods.R [hunk 1]
```diff
@@ -37,6 +37,17 @@ dkde1d <- function(x, obj) {
         x <- x[[1]]
     if (!is.ordered(x))
         stopifnot(!is.factor(x))
+    check_boundary_violations(x, obj$xmin, obj$xmax)
```
Comment (tvatter) [inline R/kde1d_methods.R]: Shouldn't it rather assign density zero outside the boundary?

File: R/kde1d_methods.R [hunk 2]
```diff
@@ -37,6 +37,17 @@ dkde1d <- function(x, obj) {
...
+
+    x <- expand_as_numeric(x)
```
Comment (tvatter) [inline R/kde1d_methods.R]: why?

File: R/kde1d_methods.R [hunk 3]
```diff
@@ -59,6 +70,7 @@ pkde1d <- function(q, obj) {
         q <- q[[1]]
     if (!is.ordered(q))
         stopifnot(!is.factor(q))
+    check_boundary_violations(q, obj$xmin, obj$xmax)
```
Comment (tvatter) [inline R/kde1d_methods.R]: Shouldn't it be 0 for elements of q <= obj$xmin and 1 for elements >= obj$xmax?

File: tests/testthat/test_kde1d.R [hunk 4]
```diff
@@ -73,10 +73,19 @@ test_that("d/p/r/h functions work", {
...
+
+        if (!is.nan(fit$xmax)) {
+            expect_error(dkde1d(xmax + 1, fit))
+            expect_error(pkde1d(xmax + 1, fit))
+        }
     }
 
     sims <- lapply(fits, function(x) rkde1d(n, x))
-    mapply(test_dpqr, fits, sims)
+    mapply(test_dpqr, fits[2], sims[2])
```
Comment (tvatter) [inline tests/testthat/test_kde1d.R]: Why?

**Review by tvatter — COMMENTED** [review]
Most of it OK, a few things that I don't get.

File: R/kde1d_methods.R (hunk 1 above)
Comment (tnagler) [reply R/kde1d_methods.R]: you're right that's better.

File: R/kde1d_methods.R (hunk 2 above)
Comment (tnagler) [reply R/kde1d_methods.R]: left-over from before your PR

File: tests/testthat/test_kde1d.R (hunk 4 above)
Comment (tnagler) [reply tests/testthat/test_kde1d.R]: left-over from bug hunting

File: R/kde1d_methods.R (hunk 3 above)
Comment (tnagler) [reply R/kde1d_methods.R]: agreed

**Pull request merged**